### PR TITLE
Added test coverage for number of retries

### LINKED ISSUE #3: Add check for number of executions in test case
_Originally posted by @ben-z in https://github.com/WATonomous/terraform-provider-email/pull/2#discussion_r1619652087_

### PR DESCRIPTION
**Pull request by AesthicEthics**
Added tests to make sure we're retrying `maxRetries` number of times.

Resolves #3

### CODE REVIEW

File: email/provider_test.go [hunk 1]
```diff
@@ -12,6 +12,9 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+// Global Testing Variables
+var testRetries int = 0
```
Comment (ben-z) [inline email/provider_test.go]:
```suggestion
var sendMailInvocations int = 0
```

Very nit, but "retry" seems to imply that there was already a "try"...

File: email/provider_test.go [hunk 2]
```diff
@@ -61,16 +64,19 @@ func testAccCheckEmailExists(n string) resource.TestCheckFunc {
 	}
 }
 func mockSendMailReturn421(addr string, a smtp.Auth, from string, to []string, msg []byte) error {
+	testRetries += 1
 	return errors.New("421 Service not available")
 }
 
 // function to test non-421 error
 func mockReturn500(addr string, a smtp.Auth, from string, to []string, msg []byte) error {
```
Comment (ben-z) [inline email/provider_test.go]:
```suggestion
func mockSendMailReturn500(addr string, a smtp.Auth, from string, to []string, msg []byte) error {
```

Didn't catch this before

File: email/provider_test.go [hunk 3]
```diff
@@ -79,34 +85,41 @@ func TestRetryWorkflow(t *testing.T) {
...
+		expectedCount int
 	}{
 		{
-			name:         "Retry with 421 error",
-			sendMailFunc: mockSendMailReturn421,
-			expectedErr:  errors.New("421 Service not available"),
+			name:          "Retry with 421 error",
+			sendMailFunc:  mockSendMailReturn421,
+			expectedErr:   errors.New("421 Service not available"),
+			expectedCount: maxRetries,
```
Comment (ben-z) [inline email/provider_test.go]:
```suggestion
			expectedSendMailInvocations: maxRetries,
```

**Review by ben-z — APPROVED** [review]
LGTM! Just nit on naming

File: email/provider_test.go [hunk 4]
```diff
@@ -79,34 +85,41 @@ func TestRetryWorkflow(t *testing.T) {
...
 	for _, test := range tests {
 		// run subtests
 		t.Run(test.name, func(t *testing.T) {
+			// set test retries zero before each test
+			testRetries = 0
 			err := sendMail(test.sendMailFunc, maxRetries, "localhost", "2525", "username", "password", "from@example.com", "to@example.com", "message")
 			if test.expectedErr != nil {
 				// assert that the errors are equal
 				assert.EqualError(t, err, test.expectedErr.Error())
+				assert.Equal(t, testRetries, test.expectedCount)
```
Comment (ben-z) [inline email/provider_test.go]: We can check this when there are no errors too

**Comment by ben-z** [comment]
I just noticed that we don’t run the tests in CI LOL. Can we add a CI job/workflow to run tests? Can be done in a separate PR

File: email/provider_test.go (hunk 4 above)
Comment (AesthicEthics) [reply email/provider_test.go]: whoops, not sure how I missed that. Thank you

**Review by ben-z — APPROVED** [review]
LGTM!

**Pull request merged**